File: core/field_vector.go
package core

import (
	"context"
	"fmt"

	validation "github.com/go-ozzo/ozzo-validation/v4"
	"github.com/pocketbase/pocketbase/core/validators"
	"github.com/pocketbase/pocketbase/tools/types"
)

func init() {
	Fields[FieldTypeVector] = func() Field {
		return &VectorField{}
	}
}

const FieldTypeVector = "vector"

// 向量索引类型常量
const (
	VectorIndexHNSW    = "hnsw"    // HNSW 索引 (推荐，更快的查询)
	VectorIndexIVFFlat = "ivfflat" // IVFFlat 索引 (更快的构建)
)

// 向量距离函数常量
const (
	VectorDistanceL2     = "l2"     // 欧几里得距离 (L2)
	VectorDistanceCosine = "cosine" // 余弦距离
	VectorDistanceIP     = "ip"     // 内积 (负内积)
)

// 最大向量维度限制
const MaxVectorDimension = 16000

var (
	_ Field = (*VectorField)(nil)
)

// VectorField 定义 "vector" 类型字段，用于存储向量嵌入。
type VectorField struct {
	Name         string `form:"name" json:"name"`
	Id           string `form:"id" json:"id"`
	System       bool   `form:"system" json:"system"`
	Hidden       bool   `form:"hidden" json:"hidden"`
	Presentable  bool   `form:"presentable" json:"presentable"`
	Dimension    int    `form:"dimension" json:"dimension"`
	Required     bool   `form:"required" json:"required"`
	IndexType    string `form:"indexType" json:"indexType"`
	DistanceFunc string `form:"distanceFunc" json:"distanceFunc"`
}

func (f *VectorField) Type() string                 { return FieldTypeVector }
func (f *VectorField) GetId() string                { return f.Id }
func (f *VectorField) SetId(id string)              { f.Id = id }
func (f *VectorField) GetName() string              { return f.Name }
func (f *VectorField) SetName(name string)          { f.Name = name }
func (f *VectorField) GetSystem() bool              { return f.System }
func (f *VectorField) SetSystem(system bool)        { f.System = system }
func (f *VectorField) GetHidden() bool              { return f.Hidden }
func (f *VectorField) SetHidden(hidden bool)        { f.Hidden = hidden }

func (f *VectorField) ColumnType(app App) string {
	return "JSON DEFAULT '[]' NOT NULL"
}

func (f *VectorField) ColumnTypePostgres() string {
	if f.Dimension > 0 {
		return fmt.Sprintf("vector(%d)", f.Dimension)
	}
	return "vector"
}

func (f *VectorField) PrepareValue(record *Record, raw any) (any, error) {
	vec := types.Vector{}
	err := vec.Scan(raw)
	return vec, err
}

func (f *VectorField) ValidateValue(ctx context.Context, app App, record *Record) error {
	val, ok := record.GetRaw(f.Name).(types.Vector)
	if !ok {
		return validators.ErrUnsupportedValueType
	}

	if val.IsZero() {
		if f.Required {
			return validation.ErrRequired
		}
		return nil
	}

	if f.Dimension > 0 && val.Dimension() != f.Dimension {
		return validation.NewError(
			"validation_vector_dimension_mismatch",
			"向量维度必须是 {{.dimension}}，实际是 {{.actual}}",
		).SetParams(map[string]any{"dimension": f.Dimension, "actual": val.Dimension()})
	}

	return nil
}

func (f *VectorField) ValidateSettings(ctx context.Context, app App, collection *Collection) error {
	return validation.ValidateStruct(f,
		validation.Field(&f.Id, validation.By(DefaultFieldIdValidationRule)),
		validation.Field(&f.Name, validation.By(DefaultFieldNameValidationRule)),
		validation.Field(&f.Dimension, validation.Min(0), validation.Max(MaxVectorDimension)),
		validation.Field(&f.IndexType, validation.In("", VectorIndexHNSW, VectorIndexIVFFlat)),
		validation.Field(&f.DistanceFunc, validation.In("", VectorDistanceL2, VectorDistanceCosine, VectorDistanceIP)),
	)
}

// CreateVectorExtensionSQL 返回创建 pgvector 扩展的 SQL。
func CreateVectorExtensionSQL() string {
	return "CREATE EXTENSION IF NOT EXISTS vector"
}

// CreateIndexSQL 返回创建向量索引的 SQL。
func (f *VectorField) CreateIndexSQL(tableName string) string {
	if f.IndexType == "" {
		return ""
	}

	indexName := fmt.Sprintf("idx_%s_%s_%s", tableName, f.Name, f.IndexType)
	opsClass := f.getOpsClass()

	switch f.IndexType {
	case VectorIndexHNSW:
		return fmt.Sprintf(
			"CREATE INDEX IF NOT EXISTS %s ON %s USING hnsw (%s %s)",
			indexName, tableName, f.Name, opsClass,
		)
	case VectorIndexIVFFlat:
		return fmt.Sprintf(
			"CREATE INDEX IF NOT EXISTS %s ON %s USING ivfflat (%s %s) WITH (lists = 100)",
			indexName, tableName, f.Name, opsClass,
		)
	default:
		return ""
	}
}

func (f *VectorField) DropIndexSQL(tableName string) string {
	if f.IndexType == "" {
		return ""
	}
	indexName := fmt.Sprintf("idx_%s_%s_%s", tableName, f.Name, f.IndexType)
	return fmt.Sprintf("DROP INDEX IF EXISTS %s", indexName)
}

func (f *VectorField) getOpsClass() string {
	switch f.DistanceFunc {
	case VectorDistanceCosine:
		return "vector_cosine_ops"
	case VectorDistanceIP:
		return "vector_ip_ops"
	default:
		return "vector_l2_ops"
	}
}

// VectorDistanceExpr 生成向量距离表达式。
func VectorDistanceExpr(column string, queryVector types.Vector, distanceFunc string) string {
	vecStr := queryVector.ToPgVector()
	switch distanceFunc {
	case VectorDistanceCosine:
		return fmt.Sprintf("%s <=> '%s'", column, vecStr)
	case VectorDistanceIP:
		return fmt.Sprintf("%s <#> '%s'", column, vecStr)
	default:
		return fmt.Sprintf("%s <-> '%s'", column, vecStr)
	}
}

File: pocketless/src/core/field_vector.ts
/**
 * VectorField — 向量字段
 * 与 Go 版 core/field_vector.go 对齐
 * SQLite: JSON 数组, PostgreSQL: pgvector VECTOR(dim)
 */

import { registerField, type Field } from "./field";
import type { RecordModel } from "./record_model";
import type { CollectionModel } from "./collection_model";

const MAX_VECTOR_DIMENSION = 16000;

export const VECTOR_INDEX_HNSW = "hnsw";
export const VECTOR_INDEX_IVFFLAT = "ivfflat";
export const VECTOR_DISTANCE_L2 = "l2";
export const VECTOR_DISTANCE_COSINE = "cosine";
export const VECTOR_DISTANCE_IP = "ip";

export class VectorField implements Field {
  id: string;
  name: string;
  type = "vector";
  system: boolean;
  hidden: boolean;
  required: boolean;
  dimension: number;
  indexType: string;
  distanceFunc: string;

  constructor(options: Record<string, unknown> = {}) {
    this.id = (options.id as string) || "";
    this.name = (options.name as string) || "";
    this.system = !!(options.system);
    this.hidden = !!(options.hidden);
    this.required = !!(options.required);
    this.dimension = (options.dimension as number) || 0;
    this.indexType = (options.indexType as string) || "";
    this.distanceFunc = (options.distanceFunc as string) || "";
  }

  columnType(isPostgres?: boolean): string {
    if (isPostgres) {
      return this.dimension > 0 ? `vector(${this.dimension})` : "vector";
    }
    return "JSON DEFAULT '[]' NOT NULL";
  }

  prepareValue(raw: unknown): number[] {
    if (raw === null || raw === undefined) return [];
    if (Array.isArray(raw)) return raw.map(Number).filter((n) => isFinite(n));
    if (typeof raw === "string") {
      try {
        const parsed = JSON.parse(raw);
        if (Array.isArray(parsed)) return parsed.map(Number).filter((n) => isFinite(n));
      } catch {}
      // pgvector "[1,2,3]" 格式
      if (raw.startsWith("[") && raw.endsWith("]")) {
        try {
          return raw.slice(1, -1).split(",").map(Number).filter((n) => isFinite(n));
        } catch {}
      }
    }
    return [];
  }

  validateValue(value: unknown, _record: RecordModel): string | null {
    const vec = Array.isArray(value) ? value : [];
    if (vec.length === 0 && this.required) return "不能为空";
    if (vec.length === 0) return null;
    if (this.dimension > 0 && vec.length !== this.dimension) {
      return `维度不匹配（期望 ${this.dimension}，实际 ${vec.length}）`;
    }
    if (vec.length > MAX_VECTOR_DIMENSION) return `维度超过上限 ${MAX_VECTOR_DIMENSION}`;
    for (const v of vec) {
      if (typeof v !== "number" || !isFinite(v)) return "包含无效数值";
    }
    return null;
  }

  validateSettings(_collection: CollectionModel): string | null {
    if (this.dimension < 0) return "维度不能为负数";
    if (this.dimension > MAX_VECTOR_DIMENSION) return `维度不能超过 ${MAX_VECTOR_DIMENSION}`;
    return null;
  }

  driverValue(value: unknown): string {
    const vec = Array.isArray(value) ? value : [];
    return JSON.stringify(vec);
  }
}

registerField("vector", (opts) => new VectorField(opts));
